File: ops/f8949.py
#!/bin/python
import sys
from utils import *

debug=False
year=18

########################################
# Read data from input files

src_dir=sys.argv[1]
build_dir=sys.argv[2]
data_dir=sys.argv[3]

personal = loadJSON(src_dir+'/personal.json')
tx_history = loadCSV(build_dir+'/tx-history.csv')

########################################
# Calculate Capital Gains/Losses

def assetsToString(assets):
  totals = {}
  msg = ""
  for asset in assets:
    totals[asset] = 0
    for lump in assets[asset]:
      totals[asset] += lump["quantity"]
    msg += " [" + asset + " " + str(totals[asset]) + "] "
  return(msg)

def chunks(l, n):
  """Yield successive n-sized chunks from l."""
  for i in range(0, len(l), n):
    yield l[i:i + n]

total_proceeds = 0
total_cost = 0
total_profit = 0
trades = []
starting_assets = assetsToString(personal['assets'])
assets = personal['assets']

for row in tx_history:

  if row['timestamp'][:2] != str(year):
    print("invalid row")
    continue

  if debug:
    print('Assets: ' + assetsToString(assets))

  if row['from'] == "ex-coinbase" or row['from'][:6] == "entity":

    if debug and row['from'][:6] == "entity":
      print('Received {} {} from {} at {} on {}'.format(row['quantity'], row['asset'], row['from'],  row['price'], row['timestamp']))
    elif debug:
      print('Bought {} {} at {} on {}'.format(row['quantity'], row['asset'], row['price'], row['timestamp']))

    if row['asset'] not in assets:
      assets[row['asset']] = []
    assets[row['asset']].append({
      "quantity": float(row['quantity']),
      "price": float(row['price'])
    })

  elif row['to'] == "ex-coinbase" or row['to'][:6] == "entity":

    if debug and row['to'][:6] == "entity":
      print('Sent {} {} to {} at {} on {}'.format(row['quantity'], row['asset'], row['to'], row['price'], row['timestamp']))
    elif debug:
      print('Sold {} {} at {} on {}'.format(row['quantity'], row['asset'], row['price'], row['timestamp']))

    profit = 0
    cost = 0

    amt = float(row['quantity'])
    while True:
      if amt == 0:
        break
      asset = assets[row['asset']][0]
      # sell off part of an asset
      if asset['quantity'] > amt:
        profit += ( (amt * float(row['price'])) - (amt * asset['price']) )
        cost += asset['price'] * amt
        asset['quantity'] -= amt
        break
      # sell off an entire asset
      else:
        profit += (float(row['price']) - asset['price']) * asset['quantity']
        cost += asset['price'] * asset['quantity']
        amt -= asset['quantity']
        assets[row['asset']].pop(0)

    proceeds = round(float(row['price']) * float(row['quantity']), 2)
    dateSold = row['timestamp'][2:4] + '/' + row['timestamp'][4:6] + '/' + row['timestamp'][0:2] 

    trades.append({
      'Description': '%s %s'  % (round(float(row['quantity']), 3), row['asset']),
      'DateAcquired': 'VARIOUS',
      'DateSold': dateSold,
      'Proceeds': proceeds,
      'Cost': round(cost, 2),
      'Code': '',
      'Adjustment': '',
      'GainOrLoss': round(profit, 2)
    })

    total_proceeds += float(row['quantity']) * float(row['price'])
    total_cost += cost
    total_profit += profit

  elif debug:
    print('idk what to do with tx from %s to %s' % (row['from'], row['to']))

########################################
# Print results of our calculations

print("\nStarting Assets:" + starting_assets)
for trade in trades:
  print('Sold %s\ton %s\tfor  %s\tPurchased for %s\t= profit of %s' % (
    trade['Description'], trade['DateSold'], trade['Proceeds'], trade['Cost'], trade['GainOrLoss']
  ))
print('Assets leftover:' + assetsToString(assets))
print('\nTotals:\t proceeds: {}\t cost: {}\t profit: {}\n'.format(round(total_proceeds,2), round(total_cost,2), round(total_profit,2)))

########################################
# Output trades in f8949 format

def buildF8949(fourteenTrades):
  f8949_data = {}
  f8949_data['FullNamePage1'] = '%s %s %s' % (
    personal['FirstName'], personal['MiddleInitial'], personal['LastName']
  )
  f8949_data['SocialSecurityNumberPage1'] = personal['SocialSecurityNumber']

  f8949_data['isShortTermA'] = False
  f8949_data['isShortTermB'] = False
  f8949_data['isShortTermC'] = True # I don't have a Form 1099-B for my crypto trades
  f8949_data['isLongTermD'] = False
  f8949_data['isLongTermE'] = False
  f8949_data['isLongTermF'] = False

  totals = { 'Proceeds': 0, 'Cost': 0, 'GainOrLoss': 0 }

  for i, trade in enumerate(fourteenTrades):
    i+=1
    totals['Proceeds'] += trade['Proceeds']
    totals['Cost'] += trade['Cost']
    totals['GainOrLoss'] += trade['GainOrLoss']
    for key, value in trade.iteritems():
      field='ST' + str(i) + key
      f8949_data[field] = str(value)

  f8949_data['STTotalProceeds'] = str(round(totals['Proceeds'], 2))
  f8949_data['STTotalCost'] = str(round(totals['Cost'], 2))
  f8949_data['STTotalAdjustment'] = ''
  f8949_data['STTotalGainOrLoss'] = str(round(totals['GainOrLoss'], 2))

  return(f8949_data)

for i, tradesChunk in enumerate(chunks(trades, 14)):
  i+=1
  outfile=data_dir+'/f8949_'+str(i)+'.json'
  print('writing to output file: ' + outfile)
  with open(outfile, "wb") as output:
    json.dump(buildF8949(tradesChunk), output)

File: src/forms/f8949.js
const f8949Mappings = require('../mappings/f8949.json');
const { add, eq, gt, lt, mul, round, sub, emptyForm, mergeForms, parseHistory } = require('../utils');

const debugMode = false;

const stringifyAssets = (assets) => {
  let output = '[\n'
  for (const [key, value] of Object.entries(assets)) {
    let total = "0"
    output += `  ${key}:`
    for (const chunk of value) {
      output += ` ${chunk.quantity}@${chunk.price},`
      total = add(total, chunk.quantity)
    }
    output += ` (Total: ${total})\n`
  }
  return `${output}]`
}

const parseF8949 = (input, output)  => {
  const f8949 = mergeForms(emptyForm(f8949Mappings), input.f8949 || {});
  const txHistory = parseHistory(input);

  // Set values constant across all f8949 forms
  f8949.FullNamePage1 = `${input.FirstName} ${input.MiddleInitial} ${input.LastName}`;
  f8949.SocialSecurityNumberPage1 = input.SocialSecurityNumber;

  const assets = input.assets || {};
  const startingAssets = stringifyAssets(assets);
  const trades = []
  let totalCost = "0"
  let totalProceeds = "0"
  let totalProfit = "0"

  debugMode && console.log(`Assets: ${stringifyAssets(assets)}`);

  for (const tx of txHistory) {
    if (!tx.timestamp.startsWith(input.taxYear.substring(2))) {
      debugMode && console.log(`Skipping old trade from ${tx.timestamp}`);
      continue
    }
    if (eq(tx.quantity, "0")) {
      debugMode && console.log(`Skipping zero-value trade of ${tx.asset} from ${tx.from} to ${tx.to}`);
      continue
    }

    if (tx.from.substring(0, 4) !== "self") {
      tx.from.substring(0, 2) === "ex"
        ? (debugMode && console.log(`Bought ${tx.quantity} ${tx.asset} from ${tx.from}`))
        : (debugMode && console.log(`Received ${tx.quantity} ${tx.asset} from ${tx.from}`))

      if (!assets[tx.asset]) {
        debugMode && console.log(`Creating new asset category for ${tx.asset}`);
        assets[tx.asset] = [];
      }

      assets[tx.asset].push({
        quantity: tx.quantity,
        price: tx.price,
      });

    } else if (tx.to.substring(0, 4) !== "self") {
      tx.to.substring(0, 2) === "ex"
        ? (debugMode && console.log(`Sold ${tx.quantity} ${tx.asset} to ${tx.to}`))
        : (debugMode && console.log(`Sent ${tx.quantity} ${tx.asset} to ${tx.to}`))


      profit = "0"
      cost = "0"

      amt = tx.quantity
      while (true) {
        if (eq(amt, "0") || lt(amt, "0")) {
          break
        }
        const asset = assets[tx.asset].pop()
        if (!asset) {
          throw new Error(`Attempting to sell more ${tx.asset} than we bought. ${tx.asset} left: ${JSON.stringify(assets[tx.asset])}`);
        }
        if (eq(asset.quantity, amt)) {
          profit = add(profit, sub(mul(amt, tx.price), mul(amt, asset.price)));
          cost = add(cost, mul(asset.price, amt));
          asset.quantity = sub(asset.quantity, amt);
          break
        } else if (gt(asset.quantity, amt)) {
          profit = add(profit, sub(mul(amt, tx.price), mul(amt, asset.price)));
          cost = add(cost, mul(asset.price, amt));
          asset.quantity = sub(asset.quantity, amt);
          assets[tx.asset].unshift(asset);
          break
        } else {
          profit = add(profit, mul(sub(tx.price, asset.price), asset.quantity));
          cost = add(cost, mul(asset.price, asset.quantity));
          amt = sub(amt, asset.quantity);
        }
      }
      const proceeds = mul(tx.price, tx.quantity);

      trades.push({
        Description: `${round(tx.quantity, 2)} ${tx.asset}`,
        DateAcquired: 'VARIOUS',
        DateSold: `${tx.timestamp.substring(2,4)}/${tx.timestamp.substring(4,6)}/${tx.timestamp.substring(0,2)}`,
        Proceeds: proceeds,
        Cost: cost,
        Code: '',
        Adjustment: '',
        GainOrLoss: profit,
      });

      totalProceeds = add(totalProceeds, proceeds);
      totalCost = add(totalCost, cost);
      totalProfit = add(totalProfit, profit);

    } else {
      console.log(`idk what to do w tx of ${tx.quantity} ${tx.asset} from ${tx.from} to ${tx.to}`);
    }

  }

  ////////////////////////////////////////
  // Print Results

  console.log(`Starting Assets: ${startingAssets}`); 
  for (const trade of trades) {
    console.log(`Sold ${trade.Description} on ${trade.DateSold} for ${trade.Proceeds} (Purchased for ${trade.Cost} = profit of ${trade.GainOrLoss}`);
  }
  console.log(`Assets Leftover: ${stringifyAssets(assets)}`); 
  console.log(`Total proceeds: ${totalProceeds} - cost ${totalCost} = profit ${totalProfit}`); 

  ////////////////////////////////////////
  // Format results into forms

  const buildF8949 = (fourteenTrades) => {
    console.log(`building form from ${fourteenTrades.length} trades`);
    subF8949 = JSON.parse(JSON.stringify(f8949));

    // TODO: identify long-term capital gains
    subF8949.isShortTermA = false;
    subF8949.isShortTermB = false;
    subF8949.isShortTermC = true; // bc I don't have a Form 1099-B
    subF8949.isLongTermD = false;
    subF8949.isLongTermE = false;
    subF8949.isLongTermF = false;

    subTotal = { Proceeds: "0", Cost: "0", GainOrLoss: "0" }

    let i = 1;
    for (const trade of fourteenTrades) {
      subTotal.Proceeds = round(add(subTotal.Proceeds, trade.Proceeds), 2);
      subTotal.Cost = round(add(subTotal.Cost, trade.Cost), 2);
      subTotal.GainOrLoss = round(add(subTotal.GainOrLoss, trade.GainOrLoss), 2);
      for (const [key, value] of Object.entries(trade)) {
        subF8949[`ST${i}${key}`] = value.match(/^-?[0-9]+.?[0-9]*$/) ? round(value, 2) : value;
      }
      i += 1;
    }
    subF8949.STTotalProceeds = round(subTotal.Proceeds, 2);
    subF8949.STTotalCost = round(subTotal.Cost, 2);
    subF8949.STTotalAdjustment = '';
    subF8949.STTotalGainOrLoss = round(subTotal.GainOrLoss, 2);

    return subF8949;
  }

  // Build a series of forms from chunks of trades
  const chunkSize = 14;
  return trades.map((e,i) =>
     i % chunkSize === 0 ? trades.slice(i, i + chunkSize) : null
  ).filter(e => !!e).map(buildF8949);
}

module.exports = { parseF8949 }
